File: Algorithms/Sorting/Quick_Sort.cpp
#include <vector>
using namespace std;

int partition(vector<int>& array, int left, int right){
  int pivot = array[right];
  int ptr = left;
  for(int i = left; i < right; i++){
    if(array[i] < pivot){
      swap(array[i], array[ptr]);
        ptr++;
    }
  }

  swap(array[ptr], array[right]);
  return ptr;
}

void qs(vector<int>& array, int left, int right){
  if(right-left > 0){
    int pivot_index = partition(array, left, right);
    qs(array, left, pivot_index-1);
    qs(array, pivot_index+1, right);
  }
}

vector<int> quickSort(vector<int> array){
  qs(array, 0, array.size()-1);
  return array;
}

File: Algorithms/Sorting/Quick_Sort.java
import java.util.*;

class Quick_Sort {

  private static int partition(int[] array, int left, int right){
    int pivot = array[right];
    int ptr = left;
    for(int i = left; i < right; i++){
      if(array[i] < pivot){
        int temp = array[i];
        array[i] = array[ptr];
        array[ptr] = temp;
        ptr++;
      }
    }
    int temp = array[right];
    array[right] = array[ptr];
    array[ptr] = temp;
    return ptr;
  }
  
  private static void qs(int[] array, int left, int right){
    if(right-left > 0){
      int pivotIndex = partition(array, left, right);
      qs(array, left, pivotIndex-1);
      qs(array, pivotIndex+1, right);
    }
  }
  
  public static int[] quickSort(int[] array) {
    // Write your code here.
    qs(array, 0, array.length - 1);
    return array;
  }
}
